service/controller/clusterapi/v17/resources/awsclusterconfig/create.go

```diff
@@ -21,7 +28,130 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 		return nil
 	}
 
-	// TODO: Map Cluster -> AWSClusterConfig.
+	// ClusterID is core part of e.g. PKI initialization etc. so it must be
+	// present before proceeding further.
+	if key.ClusterID(cluster) == "" {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("provider status in cluster cr %q does not contain cluster ID", cluster.Name))
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+		return nil
+	}
+
+	var versionBundles []versionbundle.Bundle
+	{
+		req := searcher.Request{
+			ReleaseVersion: key.ClusterReleaseVersion(cluster),
+		}
+
+		res, err := r.clusterClient.Release.Searcher.Search(ctx, req)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		versionBundles = res.VersionBundles
+	}
+
+	// Get existing AWSClusterConfig or create a new one.
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding out if AWSClusterConfig %q/%q exists", cluster.Namespace, key.AWSClusterConfigName(cluster)))
+
+	presentAWSClusterConfig, err := r.g8sClient.CoreV1alpha1().AWSClusterConfigs(cluster.Namespace).Get(key.AWSClusterConfigName(cluster), metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find AWSClusterConfig %q/%q", cluster.Namespace, key.AWSClusterConfigName(cluster)))
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating AWSClusterConfig %q/%q", cluster.Namespace, key.AWSClusterConfigName(cluster)))
+
+		newAWSClusterConfig := r.constructAWSClusterConfig(cluster, versionBundles)
+
+		_, err = r.g8sClient.CoreV1alpha1().AWSClusterConfigs(cluster.Namespace).Create(&newAWSClusterConfig)
+		if errors.IsAlreadyExists(err) {
+			r.logger.LogCtx(ctx, "level", "warning", "message", fmt.Sprintf("AWSClusterConfig %q/%q already exists", newAWSClusterConfig.Namespace, newAWSClusterConfig.Name))
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created AWSClusterConfig %q/%q", newAWSClusterConfig.Namespace, newAWSClusterConfig.Name))
+		return nil
```

Comment:
```suggestion
		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
		return nil
```

Reply:
I can add this, no problem, but I don't fully understand this. Why there should be `canceling resource` log message when the resource has fully executed at this point?

This if-branch was the case where `AWSClusterConfig` didn't exist at all yet and had to be created. During construction, it's Spec was filled and at this point the resource has done what it needs to do. After the if the rest of the function deals with update case where the object already existed. 

Initially I wrote the function in different way (https://github.com/giantswarm/cluster-operator/pull/482/commits/d27acb960c56e183bd1f372ed8c306f4c52d7652#diff-8eaeecc39fe54e3d091704d8a52bcfdfR27) because I found this present way a bit confusing, but per review comments I changed that because relying on `Generation` was judged as too obscure.

Reply:
> I can add this, no problem, but I don't fully understand this. Why there should be `canceling resource` log message when the resource has fully executed at this point?

My understanding is that the resource is fully executed in line 98 (at the time of writing this comment). I.e. lines:

```
	return nil
}
```

This here is kind of short branch (even if it isn't super short).

In previous implementation I would expect lines in this order:

```
	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updating AWSClusterConfig %q/%q", awsClusterConfig.Namespace, awsClusterConfig.Name))
	if awsClusterConfig.Generation > 0 {
...
} else {
	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not update AWSClusterConfig %q/%q; it is freshly created", awsClusterConfig.Namespace, awsClusterConfig.Name))
}
```

But I don't want to spend more time on it. This is not that important (other than I missed the `return nil` which is crucial detail in the implementation). Still this is all temporary state. Let's not spend more time on this thing.

Reply:
TL;TR I can't tell you with 100% confidence I'm right here. As you are implementor do the thing you consider more appropriate.